diff --git a/bindings/go/BUILD.bazel b/bindings/go/BUILD.bazel
index 7f2a27a..a30a3c3 100644
--- a/bindings/go/BUILD.bazel
+++ b/bindings/go/BUILD.bazel
@@ -5,7 +5,7 @@ go_library(
     name = "go",
     srcs = ["binding.go"],
     cgo = True,
-    copts = ["-std=c11 -fPIC -Ibindings/go/external/gazelle++go_deps+com_github_wharflab_tree_sitter_powershell/src"],
+    cdeps = ["//src:parser"],
     importpath = "github.com/wharflab/tree-sitter-powershell/bindings/go",
     visibility = ["//visibility:public"],
 )
diff --git a/bindings/go/binding.go b/bindings/go/binding.go
index d1ec27f..7822942 100644
--- a/bindings/go/binding.go
+++ b/bindings/go/binding.go
@@ -1,10 +1,8 @@
 package tree_sitter_powershell

-// #cgo CFLAGS: -std=c11 -fPIC -I${SRCDIR}/../../src
-// #include "../../src/parser.c"
-// #if __has_include("../../src/scanner.c")
-// #include "../../src/scanner.c"
-// #endif
+// #include "tree_sitter/parser.h"
+//
+// const TSLanguage *tree_sitter_powershell(void);
 import "C"

 import "unsafe"
diff --git a/src/BUILD.bazel b/src/BUILD.bazel
new file mode 100644
index 0000000..8a411ff
--- /dev/null
+++ b/src/BUILD.bazel
@@ -0,0 +1,11 @@
+load("@rules_cc//cc:defs.bzl", "cc_library")
+
+cc_library(
+    name = "parser",
+    srcs = ["parser.c", "scanner.c"],
+    hdrs = glob(["tree_sitter/*.h"]),
+    copts = ["-std=c11"],
+    includes = ["."],
+    visibility = ["//visibility:public"],
+)
+
