# hadolint/DL3011

Valid UNIX ports range from 0 to 65535.

| Property | Value |
|----------|-------|
| Severity | Error |
| Category | Correctness |
| Default | Enabled |

## Description

Valid UNIX ports range from 0 to 65535. Exposing a port outside this range is invalid and will result in an error.

## Examples

### Problematic code

```dockerfile
FROM busybox
EXPOSE 80000
```

### Correct code

```dockerfile
FROM busybox
EXPOSE 65535
```

## Reference

- <https://github.com/hadolint/hadolint/wiki/DL3011>
